javabuilder-authorizer/lambda_function.rb

```diff
@@ -0,0 +1,88 @@
+require 'aws-sdk-lambda'
+require 'jwt'
+
+# The lambda handler takes an event with the header
+# 'Authorization: Bearer <token>',  where the token is a JWT token
+# generated by dashboard. The handler verifies the token with the appropriate
+# public key (dependant on the environment the request came from),
+# and checks the token has not expired and its issue time is not in the future.
+def lambda_handler(event:, context:)
+  headers = event['headers']
+  auth_header = headers['Authorization']
+  method_arn = event['methodArn']
+
+  if auth_header.start_with?('Bearer: ')
+    jwt_token = auth_header.split[1]
+    begin
+      decoded_token = JWT.decode(
+        jwt_token,
+        get_public_key(method_arn),
+        true,
+        verify_iat: true, # verify issued at time is valid
+        algorithm: 'RS256'
+      )
+    rescue JWT::ExpiredSignature, JWT::InvalidIatError
+      return generate_deny(nil, method_arn)
+    end
+
+    return generate_deny(nil, method_arn) unless decoded_token
+
+    user_id = decoded_token[0]['uid']
+    generate_allow(user_id, method_arn, decoded_token[0])
```

Comment:
Since the `generate_allow` method invokes the `generate_policy` method, passing the `user_id` as the `principal_id` in the policy document, I think we would need to configure every Code Studio user in AWS IAM as a `Principal` before we receive the HTTP request? I think we may need to statically provision an IAM Role for a javalab client via CloudFormation and then have the authorizer allow the request to assume that role? I’ll research tomorrow as well.

Reply:
ah ok good to know. I couldn't find any information on what the principal id needed to be, and it just worked with both 'me' and a random user id so I didn't look much further. Let me know what is recommended here.

Reply:
How could this have worked for Molly's local testing? I don't think any of our Code Studio users are set up in AWS IAM.

Reply:
I'm still researching this morning on the best way to set the generated policy (and the principal_id specified in the policy).

Reply:
I'm used to a Principal identifying a specific User or Role defined in AWS IAM or an AWS Service that needs to invoke other AWS APIs on our behalf, but [reading discussions online](https://forums.aws.amazon.com/thread.jspa?threadID=225671), it's clear that setting this to the Code Studio user id is correct and we don't need to provision those users ahead of time anywhere in AWS. The user id will get logged as the Principal invoking the API Gateway route in some places, so might be useful to prefix/namespace the user ids: `studio.code.org/user_id/####`, `staging-studio.code.org/user_id/####`, etc.

Reply:
awesome, thank you! Should the user id we pass in the token be namespaced or should just the principal id be namespaced?

Reply:
prefixing the user_id with the `issuer` from the token might work. We'd have to update the Code Studio controller that issues the token to set the `issuer` to the Code Studio dashboard domain name (`CDO.dashboard_domain` ?)

Reply:
added issuer [here](https://github.com/code-dot-org/code-dot-org/pull/39992) and updated the principal id to be namespaced with the issuer.